### DISCUSSION TITLE
uuid handling differences between 1.x and 2.x

### DISCUSSION THREAD

**Discussion by twillis** [sqlalchemy/sqlalchemy#12158] [Usage Questions]
Hello, Im in the process(finally)of upgrading from 1.4. and the biggest issue at the moment, is the difference in how uuids are handled.

this is for a pg database all primary/foreign keys are of type uuid.

as far as I can tell, in 1.X a string passed that contains a uuid will get into the sql correctly and if you pass a type UUID it will work.

on 2.x it is now finicky with UUID vs. strings. compares dont work of course, and the json serialization we have is busted etc.

I can understand the change in that yeah, its probably better to just use uuids and not strings, for type safety and all that, but is there a way for me to turn on the 1.4 behavior so I can at least get things upgraded before correcting the code bases UUID sins?

after reading the docs I came up with this and it at least allows for things like

instance.event_uid = str(uuid.uuid4()) 

Should I use Uuid instead? 

Am i screwed here?

**Comment by zzzeek** [comment]
Lots of things have gone on with uuids but I'm not familiar with which specific change this regards.  The UUID datatype has always been one that works with only one Python datatype per UUID object, that is, only Python `uuid` objects if you have `UUID(as_uuid=True)` and only strings if you have `UUID(as_uuid=False)`, however at the same time you can pass string UUIDs into it and still get results if the datatype has as_uuid turned on, though this might have some dependency on driver.

Try out the program below, it round trips UUIDs and illustrates that comparing to string or uuid works for the PostgreSQL psycopg2 driver which is what you'd have in 1.4, and the program delivers the identical results in 1.4 or 2.0.  So I would need specifics on what the problem is.

```py
import uuid

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(UUID(as_uuid=True))

e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

s = Session(e)

uuid_obj = uuid.uuid4()
string_uuid = str(uuid_obj)

# insert data as a string, even though it's as_uuid, succeeds
s.add(A(data=string_uuid))
s.commit()

# data was persisted
assert s.query(A.data).scalar() == uuid_obj

# use WHERE clasue comparing to uuid object, succeeds
assert s.query(A.data).where(A.data == uuid_obj).scalar() == uuid_obj

# use WHERE clasue comparing to string, succeeds
assert s.query(A.data).where(A.data == string_uuid).scalar() == uuid_obj
```

**Reply by twillis** [reply]
ok, i'll dig deeper into this. thx for the response.

**Reply by CaselIT** [reply]
what driver are you using? psycopg2?
